### PR TITLE
FEATURE #14: Apply filters to light data frames

### PR DESCRIPTION
**Pull request by warut-vijit**
Adds filters to the YAML file and creates the Filter class. Also adds sample filter, greyscale

### CODE REVIEW

File: filters/greyscale/__init__.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                unit_pixels = []
+                for pixel in tile_matrix[unit]:
+                    avg = sum(pixel)/3
+                    unit_pixels.append((avg, avg, avg))
+                output_matrix[unit] = unit_pixels
+        self.last_frame_time = time.time()
+        return output_matrix
+
+
+filter = GreyscaleFilter
```
Comment (wchill) [inline filters/greyscale/__init__.py]: filter is a keyword in Python, we should probably actually change the terminology

File: filter_base.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+    def active(self):
+        """Defines whether this filter will affect the current frame"""
+        return False
+
+    def applyToFrame(self, tile_matrix):
+        """Receives as input one frame of color data in a dictionary
+        where unit IDs map to either a tuple or a list of tuples. Returns
+        filtered color data in the same format."""
+        raise NotImplementedError
```
Comment (wchill) [inline filter_base.py]: Filters should also be able to drop frames if necessary (by returning None, perhaps)

File: plugin_manager.py [hunk 3]
```diff
@@ -13,40 +13,69 @@ class PluginManager:
...
+                available_filters[filter_folder] = {
+                    "name": filter_folder,
+                    "path": location+filter_folder+"/__init__.py",
+                    "priority": 0
+                }
+        return available_filters
+
     def load_plugin(self, plugin_name):
-        """Loads plugin named plugin_name from plugins folder"""
+        """Loads plugin named plugin_name from plugins folder. Returns 0 if successful, else 1"""
```
Comment (wchill) [inline plugin_manager.py]: Why not just return True/False, this isn't C

File: plugin_manager.py [hunk 4]
```diff
@@ -59,11 +88,28 @@ def load_plugin(self, plugin_name):
...
+                    return 0
             self.loaded_plugins.append(plugin_obj)
-        else:
-            # plugin not found
-            pass
+            return 0
+        return 1
+
+    def load_filter(self, filter_name):
+        """Loads filter named filter_name from filters folder. Returns 0 if successful, else 1"""
```
Comment (wchill) [inline plugin_manager.py]: Same here

File: filter_base.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class Filter:
+
+    def __init__(self):
+        pass
+
+    def active(self):
+        """Defines whether this filter will affect the current frame"""
+        return False
+
+    def applyToFrame(self, tile_matrix):
```
Comment (modi95) [inline filter_base.py]: snek_case pls ❤️

File: filters/greyscale/__init__.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import time
+from filter_base import Filter
+
+
+class GreyscaleFilter(Filter):
```
Comment (modi95) [inline filters/greyscale/__init__.py:5]: snek_case here and elsewhere

File: plugin_manager.py [hunk 7]
```diff
@@ -13,40 +13,69 @@ class PluginManager:
...
         config_loader.register_config('plugin', PLUGIN_FILE_NAME, default_config)
         self.plugins = config_loader.load_config('plugin')
         self.loaded_plugins = []
-        self.blocked = {}  # Dictionary to prevent multiple additions of same plugin to queue
+        self.loaded_filters = []
+        self.blocked = {}  # Dictionary to prevent multiple additions of same plugin/filter to queue
         self.config = config_loader.load_config('twilight')
 
+    def get_files(self, location):
+        """Helper function to retrieve plugin and filter scripts"""
```
Comment (modi95) [inline plugin_manager.py]: Please describe args and return values. 
See PEP 257 / twilight.py for reference

Fix docstrings similarly elsewhere as well please.

File: plugin_manager.py [hunk 8]
```diff
@@ -13,40 +13,69 @@ class PluginManager:
...
+        location = "." + self.config["FILTERS_FOLDER"]
+        for filter_folder in self.get_files(location):
+            if filter_folder in self.plugins["FILTERS"]:
+                available_filters[filter_folder] = {
+                    "name": filter_folder,
+                    "path": location+filter_folder+"/__init__.py",
+                    "priority": self.plugins["FILTERS"][filter_folder]["priority"]
+                }
+            else:
+                available_filters[filter_folder] = {
```
Comment (modi95) [inline plugin_manager.py]: Remove this failsafe and throw an exception instead please.
Ping me if this seems like an odd request 😄

File: plugin_manager.py [hunk 9]
```diff
@@ -59,11 +88,28 @@ def load_plugin(self, plugin_name):
...
-                    return
+                    return 0
             self.loaded_plugins.append(plugin_obj)
-        else:
-            # plugin not found
-            pass
+            return 0
+        return 1
+
+    def load_filter(self, filter_name):
```
Comment (modi95) [inline plugin_manager.py]: This seems like quite a bit of duplication of code. Please add a TODO to figure out how we can fix this in the future.

File: plugins.yml [hunk 10]
```diff
@@ -11,3 +11,8 @@ PLUGINS:
   mood:
     priority: 0
     persistent: False
+
+# Filter data, priority in range 0-9 inclusive. Filters are persistent by default
+FILTERS:
+  flux:
+    priority: 0
```
Comment (modi95) [inline plugins.yml]: Please add newline at the end of the file.

**Review by modi95 — CHANGES_REQUESTED** [review]

**Review by wchill — CHANGES_REQUESTED** [review]

File: plugin_manager.py (hunk 3 above)
Comment (wchill) [reply plugin_manager.py]: Actually, make it raise an Exception

Comment (warut-vijit) [reply plugin_manager.py]: I'm not sure how you want the program to run, but according to what I understand, the script should continue running even if the plugin fails to load. That is, if one person makes an error in requesting a plugin, it shouldn't cause the entire Twilight system to crash.

File: plugin_manager.py (hunk 4 above)
Comment (warut-vijit) [reply plugin_manager.py]: See above

**Comment by modi95** [comment]
There are a couple of more 🐍 _case issues. Also the line ending at the end of the plugins.yml file. 

edit: 🐍 case issues are fixed. Sorry about that. Just the last line ending at plugins.yml 😄

File: plugin_manager.py (hunk 3 above)
Comment (modi95) [reply plugin_manager.py]: Depending on the behavior we want later, we can catch the exception in whatever is running this. It's basically convention.

File: plugin_manager.py [hunk 11]
```diff
@@ -89,8 +136,9 @@ def start(self):
                 # Load the next plugin and give it the current panel layout
                 current_module = next_module
                 current_plugin = imp.load_source("__init__", next_module["path"]).plugin()
```
Comment (wchill) [inline plugin_manager.py]: I forgot to mention that `imp.load_source` is deprecated. See #20 for how I solved the deprecation issue.

File: filters/greyscale/__init__.py (hunk 6 above)
Comment (modi95) [reply filters/greyscale/__init__.py:5]: Marking as done because this comment was on a class name

File: plugin_manager.py (hunk 11 above)
Comment (wchill) [reply plugin_manager.py]: https://github.com/acm-uiuc/twilight/blob/a9cbf7a0cce9b18518bcf8e0ff2a675bdfaafa01/plugin_manager.py#L94

**Review by modi95 — APPROVED** [review]

**Review by wchill — APPROVED** [review]

**Comment by wchill** [comment]
Don't merge yet until I've had a chance to test the branch

**Review by wchill — CHANGES_REQUESTED** [review]
Need to test

**Review by wchill — APPROVED** [review]

**Comment by wchill** [comment]
As I am stepping down from this project, I will no longer be able to test this PR.

**Pull request merged**